### PR TITLE
Fix fugue empty partition bug

### PR DESCRIPTION
**Pull request by goodwanghan**
Fixing https://github.com/capitalone/datacompy/issues/218

### CODE REVIEW

File: tests/test_fugue.py [hunk 1]
```diff
@@ -268,22 +290,81 @@ def test_doc_case():
...
+
+    comp = Compare(
+        no_intersection_diff_df1, no_intersection_diff_df2, join_columns=["x"]
+    )
+    a = report(no_intersection_diff_df1, no_intersection_diff_df2, ["x"])
+    _compare_report(comp.report(), a)
+    a = report(no_intersection_diff_df1, no_intersection_diff_df2, "x", parallelism=2)
+    _compare_report(comp.report(), a)
+
+    # can't pass due to https://github.com/capitalone/datacompy/issues/221
```
Comment (fdosani) [inline tests/test_fugue.py:322]: given that the issue you flagged is sort of a limitation we can maybe do something like this? appreciate your thoughts.

```python 
@pytest.fixture
def large_diff_df1():
    np.random.seed(0)
    data = np.random.randint(0, 7, size=10000)
    return pd.DataFrame({"x": data, "y": np.array([9]*10000)})


@pytest.fixture
def large_diff_df2():
    np.random.seed(0)
    data = np.random.randint(6, 11, size=10000)
    return pd.DataFrame({"x": data, "y": np.array([9]*10000)})


def test_report_pandas(
    simple_diff_df1,
    simple_diff_df2,
    no_intersection_diff_df1,
    no_intersection_diff_df2,
    large_diff_df1,
    large_diff_df2,
):
    comp = Compare(simple_diff_df1, simple_diff_df2, join_columns=["aa"])
    a = report(simple_diff_df1, simple_diff_df2, ["aa"])
    _compare_report(comp.report(), a)
    a = report(simple_diff_df1, simple_diff_df2, "aa", parallelism=2)
    _compare_report(comp.report(), a)

    comp = Compare(
        no_intersection_diff_df1, no_intersection_diff_df2, join_columns=["x"]
    )
    a = report(no_intersection_diff_df1, no_intersection_diff_df2, ["x"])
    _compare_report(comp.report(), a)
    a = report(no_intersection_diff_df1, no_intersection_diff_df2, "x", parallelism=2)
    _compare_report(comp.report(), a)

    # due to https://github.com/capitalone/datacompy/issues/221
    # we can have y as a constant to ensure all the x matches are equal

    comp = Compare(large_diff_df1, large_diff_df2, join_columns=["x"])
    a = report(large_diff_df1, large_diff_df2, ["x"])
    _compare_report(comp.report(), a, truncate=True)
    a = report(large_diff_df1, large_diff_df2, "x", parallelism=2)
    _compare_report(comp.report(), a, truncate=True)


def test_report_spark(
    spark_session,
    simple_diff_df1,
    simple_diff_df2,
    no_intersection_diff_df1,
    no_intersection_diff_df2,
    large_diff_df1,
    large_diff_df2,
):
    simple_diff_df1.iteritems = simple_diff_df1.items  # pandas 2 compatibility
    simple_diff_df2.iteritems = simple_diff_df2.items  # pandas 2 compatibility
    no_intersection_diff_df1.iteritems = (
        no_intersection_diff_df1.items
    )  # pandas 2 compatibility
    no_intersection_diff_df2.iteritems = (
        no_intersection_diff_df2.items
    )  # pandas 2 compatibility
    large_diff_df1.iteritems = large_diff_df1.items  # pandas 2 compatibility
    large_diff_df2.iteritems = large_diff_df2.items  # pandas 2 compatibility

    df1 = spark_session.createDataFrame(simple_diff_df1)
    df2 = spark_session.createDataFrame(simple_diff_df2)
    comp = Compare(simple_diff_df1, simple_diff_df2, join_columns="aa")
    a = report(df1, df2, ["aa"])
    _compare_report(comp.report(), a)

    df1 = spark_session.createDataFrame(no_intersection_diff_df1)
    df2 = spark_session.createDataFrame(no_intersection_diff_df2)
    comp = Compare(no_intersection_diff_df1, no_intersection_diff_df2, join_columns="x")
    a = report(df1, df2, ["x"])
    _compare_report(comp.report(), a)

    # due to https://github.com/capitalone/datacompy/issues/221
    # we can have y as a constant to ensure all the x matches are equal

    df1 = spark_session.createDataFrame(large_diff_df1)
    df2 = spark_session.createDataFrame(large_diff_df2)
    comp = Compare(large_diff_df1, large_diff_df2, join_columns="x")
    a = report(df1, df2, ["x"])
    _compare_report(comp.report(), a, truncate=True)
```

The key bit being `large_diff_df1` and `large_diff_df2`'s `y` just being some constant so say if there are are a random number of `x` join keys on the first `n` will match in both the fugue and non-fugue version. It should give reproducible results.

**Review by fdosani — COMMENTED** [review]
LGTM. Just a single question and also suggestion for the unit test.

File: tests/test_fugue.py [hunk 2]
```diff
@@ -89,6 +88,29 @@ def simple_diff_df2():
...
+
+@pytest.fixture
+def no_intersection_diff_df2():
+    return pd.DataFrame(dict(x=["b"], y=[1.1]))
+
+
+@pytest.fixture
+def large_diff_df1():
+    np.random.seed(0)
+    return pd.DataFrame(np.random.randint(0, 7, size=(10000, 2)), columns=["x", "y"])
```
Comment (fdosani) [inline tests/test_fugue.py:105]:
```suggestion
    data = np.random.randint(0, 7, size=10000)
    return pd.DataFrame({"x": data, "y": np.array([9]*10000)})
```

File: tests/test_fugue.py [hunk 3]
```diff
@@ -89,6 +88,29 @@ def simple_diff_df2():
...
+
+
+@pytest.fixture
+def large_diff_df2():
+    np.random.seed(0)
+    return pd.DataFrame(np.random.randint(6, 11, size=(10000, 2)), columns=["x", "y"])
```
Comment (fdosani) [inline tests/test_fugue.py:111]:
```suggestion
    data = np.random.randint(6, 11, size=10000)
    return pd.DataFrame({"x": data, "y": np.array([9]*10000)})
```

File: tests/test_fugue.py [hunk 4]
```diff
@@ -268,22 +290,81 @@ def test_doc_case():
...
+    _compare_report(comp.report(), a)
+    a = report(no_intersection_diff_df1, no_intersection_diff_df2, "x", parallelism=2)
+    _compare_report(comp.report(), a)
+
+    # can't pass due to https://github.com/capitalone/datacompy/issues/221
+    # comp = Compare(large_diff_df1, large_diff_df2, join_columns=["x"])
+    # a = report(large_diff_df1, large_diff_df2, ["x"])
+    # _compare_report(comp.report(), a, truncate=True)
+    # a = report(large_diff_df1, large_diff_df2, "x", parallelism=2)
+    # _compare_report(comp.report(), a, truncate=True)
```
Comment (fdosani) [inline tests/test_fugue.py:327]:
```suggestion
    # due to https://github.com/capitalone/datacompy/issues/221
    # we can have y as a constant to ensure all the x matches are equal

    comp = Compare(large_diff_df1, large_diff_df2, join_columns=["x"])
    a = report(large_diff_df1, large_diff_df2, ["x"])
    _compare_report(comp.report(), a, truncate=True)
    a = report(large_diff_df1, large_diff_df2, "x", parallelism=2)
    _compare_report(comp.report(), a, truncate=True)
```

File: tests/test_fugue.py [hunk 5]
```diff
@@ -268,22 +290,81 @@ def test_doc_case():
...
+    comp = Compare(no_intersection_diff_df1, no_intersection_diff_df2, join_columns="x")
+    a = report(df1, df2, ["x"])
+    _compare_report(comp.report(), a)
+
+    # can't pass due to https://github.com/capitalone/datacompy/issues/221
+    # df1 = spark_session.createDataFrame(large_diff_df1)
+    # df2 = spark_session.createDataFrame(large_diff_df2)
+    # comp = Compare(large_diff_df1, large_diff_df2, join_columns="x")
+    # a = report(df1, df2, ["x"])
+    # _compare_report(comp.report(), a, truncate=True)
```
Comment (fdosani) [inline tests/test_fugue.py:367]:
```suggestion
    # due to https://github.com/capitalone/datacompy/issues/221
    # we can have y as a constant to ensure all the x matches are equal

    df1 = spark_session.createDataFrame(large_diff_df1)
    df2 = spark_session.createDataFrame(large_diff_df2)
    comp = Compare(large_diff_df1, large_diff_df2, join_columns="x")
    a = report(df1, df2, ["x"])
    _compare_report(comp.report(), a, truncate=True)
```

**Review by fdosani — COMMENTED** [review]
Adding suggestion. Feel free to ignore or suggest something else.

**Comment by fdosani** [comment]
@goodwanghan friendly ping if you have time to review the comments.

**Comment by goodwanghan** [comment]
> @goodwanghan friendly ping if you have time to review the comments.

@fdosani my sincere apology for the delay. I am on vacation since this week. But I will make sure to address the comments by the end of next week.

File: setup.cfg [hunk 6]
```diff
@@ -35,7 +35,7 @@ install_requires =
 	pandas<=2.0.2,>=0.25.0
 	numpy<=1.24.3,>=1.22.0
 	ordered-set<=4.1.0,>=4.0.2
-	fugue<=0.8.5,>=0.8.5
+	fugue<=0.9.0,>=0.8.5
```
[the remark this answers is not in the document]
Comment (goodwanghan) [reply setup.cfg:38]: Yes, this is to protect the future breaking change. Fugue 0.9.0 will have a couple of breaking changes.

File: tests/test_fugue.py (hunk 1 above)
Comment (goodwanghan) [reply tests/test_fugue.py:322]: Thank you so much! I have replaced with your test.

**Review by fdosani — APPROVED** [review]
LGTM

**Pull request merged**